src/main.rs

```diff
@@ -131,6 +137,69 @@ struct CameraErr {
     err: neolink::Error,
 }
 
+fn set_up_tls(config: &Config, rtsp: &RtspServer) {
+    let tls_client_auth = match &config.tls_client_auth as &str {
+        "request" => TlsAuthenticationMode::Requested,
+        "require" => TlsAuthenticationMode::Required,
+        "none" => TlsAuthenticationMode::None,
+        _ => unreachable!(),
+    };
+    if let Some(cert_path) = &config.certificate {
+        rtsp.set_tls(&cert_path, tls_client_auth).expect("Failed to set up TLS");
+    }
+}
+
+fn set_up_users(users: &Vec<UserConfig>, rtsp: &RtspServer) {
+    // Setting up users
+    let mut credentials = vec![];
+    let reserved_names = vec!(String::from("anyone"), String::from("anonymous"));
+    for user in users {
+        if let Some(name) = &user.name {
+            if reserved_names.contains(name) {
+                warn!("{} is a reserved username, skipping this user", name);
+                continue;
+            }
+        }
+        let name = &user.name;
+        let pass = &user.pass;
+        let user_pass = match (name, pass)  {
+            (Some(name), Some(pass)) => {
+                Some((&name as &str, &pass as &str))
+            },
+            (Some(_), None) | (None, Some(_)) => {
+                warn!("Username and password must be supplied together - ignoring [[users]] entry");
+                None
+            },
+            _ => None,
+        };
+        credentials.push(user_pass);
+    }
+    rtsp.set_credentials(&credentials).expect("Failed to set up users.");
+}
+
+fn get_permitted_users(users: &Vec<UserConfig>, current_permitted_users: &Vec<String>) -> Vec<String> {
+    // This is required to handle the special case of "anyone"
+    // ===Special set up of "anyone"===
+    // If in the camera config there is the user "anyone"
+    // Then we add all users to the cameras config. including unauth
+    let mut new_permitted_users = vec![];
+    if current_permitted_users.contains(&"anyone".to_string()) {
+        for credentials in users {
+            if let Some(user) = &credentials.name {
+                new_permitted_users.push(user.to_string());
+            }
+        }
+        new_permitted_users.push("anonymous".to_string());
```

Comment:
Ooh so in my head "anonymous" is distinct from "anyone". Let's nail the semantics down. To me:

- `anyone` means "any properly authenticated user"
- `anonymous` means "any connecting user"

Do you agree?

Reply:
That sounds like a reasonable method. I think I should change the default to `anonymous` though so that it won't effect current configs.

Reply:
Or should we have `anonymous` if no users in the config or `anyone` when users are in the config?

Reply:
The problem with that way though is how to set up the condition in the toml default function...

Reply:
So I set it up like this, permitted_users is now an option which defaults to None.

In the `get_permitted_users` function we check for None or Some
- if Some then use what is in the toml
- if None check
  - if length of users>0  then use `[ "anyone" ]`
  - if length of users == 0 use `[ "anonymous" ]`

Reply:
So I set it up like this. `permitted_users` is now an Option which defaults to None

In the function `get_permitted_users` we match `permitted_users` to
- Some, use what was in the toll
- None, we check the length of users I.e. number of `[[users]]`
  - If length > 0 `permited_users` defaults to `[ "anyone" ]`. Meaning any of the `[[users]]` can connect but not anonymous. This means you can activate authentication with just a `[[users]]` block.
  - If length == 0 `permited_users` defaults to `[ "anoymous" ]`. Meaning anyone even without authentication. This reflects the current default so no changes in config required.

Reply:
Yep, I think that's the behavior I'd expect. Sounds good.